agent/src/main/java/com/appland/appmap/config/AppMapConfig.java

```diff
@@ -131,4 +169,81 @@ public boolean isShallow(FullyQualifiedName canonicalName) {
+    pw.format("name: %s\n", CLI.projectName(new File(directory)));
+
+    // For now, this only works in this type of standardize repo structure.
+    File javaDir = Paths.get(directory).resolve("src/main/java").toFile();
+    if (javaDir.isDirectory()) {
+      int pkgStart = javaDir.toPath().getNameCount();
+      // Collect package names in src/main/java
+      Set<Path> packages = new HashSet<>();
+      Files.walkFileTree(javaDir.toPath(), new SimpleFileVisitor<Path>() {
+        @Override
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+          if (file.getFileName().toString().endsWith(".java")) {
+            int pkgEnd = file.getParent().getNameCount();
+            if (pkgStart == pkgEnd) {
+              // We're in the the unnamed package, ignore
+              return FileVisitResult.CONTINUE;
+            }
+
+            Path packagePath = file.getParent().subpath(pkgStart, pkgEnd);
+            if (packagePath.getNameCount() > 0) {
+              packages.add(packagePath);
+            }
+          }
+          return FileVisitResult.CONTINUE;
+        }
+      });
+
+      pw.print("\n"
+          + "# Your project contains the directory src/main/java. AppMap has\n"
+          + "# auto-detected Java packages. By default, only classes in these\n"
+          + "# packages will be recorded. To record classes in subpackages, remove\n"
+          + "# the line(s)\n"
+          + "#   shallow: true\n"
+          + "\n"
+          + "packages:\n");
+
+      // Collect just the packages that don't have a matching ancestor in the package
+      // list.
+      List<Path> topLevelPackages = packages.stream().sorted().collect(ArrayList::new, (memo, packagePath) -> {
+        for (int i = 1; i < packagePath.getNameCount(); i++) {
+          Path ancestorPath = packagePath.subpath(0, i);
+          if (memo.contains(ancestorPath)) {
+            return;
+          }
+        }
+        memo.add(packagePath);
+      }, ArrayList::addAll);
+      topLevelPackages.forEach(new Consumer<Path>() {
+        @Override
+        public void accept(Path packagePath) {
+          List<String> tokens = new ArrayList<>();
+          for (int i = 0; i < packagePath.getNameCount(); i++) {
+            tokens.add(packagePath.getName(i).toString());
+          }
+          String path = String.join(".", tokens);
+          pw.format("- path: %s\n", path);
+          pw.format("  shallow: true\n");
+        }
+      });
+    } else {
+      pw.println("packages: []");
+      pw.println("# appmap-java init looks for source packages in src/main/java.");
+      pw.println("# This folder was not found in your project, so no packages were auto-detected.");
+      pw.println("# You can add your source packages by replacing the line above with lines like this:");
+      pw.println("# packages:");
+      pw.println("# - path: com.mycorp.pkg");
+      pw.println("# - path: org.otherstuff.pkg");
+    }
+    return sw;
```

Comment:
Feels like this should just return the string instead, the StringWriter being an implementation detail.